Geometry/FindIntersections.js

```diff
@@ -0,0 +1,199 @@
+ * @param {Segment} b - The second segment.
+ * @returns {number} Negative if a < b, positive if a > b, zero if equal.
+ */
+export const compareSegmentsY = (a, b) => a.start.y - b.start.y
+
+/**
+ * Calculates the intersection point of two line segments.
+ * @param {Segment} seg1 - First segment.
+ * @param {Segment} seg2 - Second segment.
+ * @returns {Point|null} The intersection point, or null if segments are parallel.
+ */
+export const calculateIntersectionPoint = (seg1, seg2) => {
+  const x1 = seg1.start.x,
+    y1 = seg1.start.y
+  const x2 = seg1.end.x,
+    y2 = seg1.end.y
+  const x3 = seg2.start.x,
+    y3 = seg2.start.y
+  const x4 = seg2.end.x,
+    y4 = seg2.end.y
+
+  const denom = (x1 - x2) * (y3 - y4) - (y1 - y2) * (x3 - x4)
+  if (Math.abs(denom) < Number.EPSILON) return null // parallel lines
+
+  const x =
+    ((x1 * y2 - y1 * x2) * (x3 - x4) - (x1 - x2) * (x3 * y4 - y3 * x4)) / denom
+  const y =
+    ((x1 * y2 - y1 * x2) * (y3 - y4) - (y1 - y2) * (x3 * y4 - y3 * x4)) / denom
+
+  return createPoint(x, y)
+}
+
+/**
+ * Handles potential intersection between two segments.
+ * @param {Segment} seg1 - First segment.
+ * @param {Segment} seg2 - Second segment.
+ * @param {Array<Event>} events - Array of events to update.
+ * @param {Array<Intersection>} intersections - Array of intersections to update.
+ */
+export const handlePotentialIntersection = (
+  seg1,
+  seg2,
+  events,
+  intersections
+) => {
+  const intersectionPoint = calculateIntersectionPoint(seg1, seg2)
+  if (intersectionPoint) {
+    events.push(createEvent(intersectionPoint, ['intersection'], [seg1, seg2]))
+    intersections.push({
+      segment1: seg1,
+      segment2: seg2,
+      point: intersectionPoint
+    })
+  }
+}
+
+/**
+ * Finds all intersections among a set of line segments using the Bentley-Ottmann algorithm.
+ * @param {Array<Segment>} segments - Array of line segments to check for intersections.
+ * @returns {Array<Intersection>} Array of all intersections found.
+ */
+const findIntersections = (segments) => {
+  const events = []
+  const intersections = []
+  let sweepLineStatus = []
+
+  // Create initial events
+  segments.forEach((segment) => {
+    events.push(createEvent(segment.start, ['left'], [segment]))
+    events.push(createEvent(segment.end, ['right'], [segment]))
+  })
+
+  // Sort events
+  events.sort((a, b) => {
+    const pointCompare = comparePoints(a.point, b.point)
+    if (pointCompare !== 0) return pointCompare
+
+    // If points are the same, prioritize: intersection > left > right
+    const typePriority = { intersection: 0, left: 1, right: 2 }
+    return (
+      Math.min(...a.types.map((t) => typePriority[t])) -
+      Math.min(...b.types.map((t) => typePriority[t]))
+    )
+  })
+
+  // Process events
+  events.forEach((event) => {
+    if (event.types.includes('left')) {
+      event.segments.forEach((segment) => {
+        sweepLineStatus.push(segment)
+        sweepLineStatus.sort(compareSegmentsY)
```

Comment:
This is O(n log n), resulting in a total time complexity of O(n² log n), which is worse than the naive brute force algorithm.

As said, the sweep line status structure needs to use a sorted set / map data structure. There should be efficient implementations of some in this repo; you need to use one.